lib/AST/ASTContext.cpp

```diff
@@ -2441,11 +2445,12 @@ bool ASTContext::canImportModule(ImportPath::Module ModuleName,
 }
 
 ModuleDecl *
-ASTContext::getModule(ImportPath::Module ModulePath) {
+ASTContext::getModule(ImportPath::Module ModulePath, bool AllowMemoryCached) {
```

Comment:
Should we not add the module to the loaded modules when `AllowMemoryCached` is set? Ie. `SerializedModuleLoader.cpp` will call `Ctx.addLoadedModule(M);` after loading the module and that will replace the already loaded one. But that means that if index module A and then module B, which depends on A, we would then have the loaded-from-interface module A when indexing B.

Or does not adding the module cause issues later, eg. something could be expecting to find that module in the loaded modules map and then getting the old B when we're indexing the loaded-from-interface B?

Mostly I'm concerned that we're not expecting this type of thing to happen in the compiler, so it's possible something is caching results that will cause problems here. I don't *know* if that's true. Though the comment about reloading the stdlib breaking references would point to something like this.

Reply:
I agree with you about the potential problem here. If the current approach works, I'll look into adding what you suggest, at least for completeness purposes and it may be more important when indexing larger projects.